code/modules/vehicles/aircraft.dm

```diff
@@ -0,0 +1,1308 @@
+	if(CHECK_BITFIELD(flags_pass, FLYING))
+		turn_while_flying(angle_to_dir(Get_Angle(source, over_object)))
+
+///Proc that grabs a new target for firing at
+/obj/vehicle/sealed/aircraft/proc/turn_gun(datum/source, atom/src_object, atom/over_object, turf/src_location, turf/over_location, src_control, over_control, params)
+	SIGNAL_HANDLER
+	if(!gun_operation_checks(over_object))
+		return FALSE
+	current_weapon.change_target(source, src_object, over_object, src_location, over_location, src_control, over_control, params)
+
+///Autofire component takes over
+/obj/vehicle/sealed/aircraft/proc/start_fire(datum/source, atom/object, turf/location, control, params)
+	SIGNAL_HANDLER
+	if(!gun_operation_checks(object))
+		return FALSE
+	current_weapon.start_fire(source, object, location, control, params, TRUE)
+
+///Code for toggling the engine on and off, easier to have it separated than as part of the action button
+/obj/vehicle/sealed/aircraft/proc/toggle_engine()
+	if(!CHECK_BITFIELD(flags_pass, FLYING) && current_fuel > 0)
+		if(!pilot)
+			return
+		if(!COOLDOWN_CHECK(src, engine_cooldown))
+			balloon_alert(pilot, "Engine is not ready!")
+			return
+		COOLDOWN_START(src, engine_cooldown, 10 SECONDS)
+		set_flying(TRUE)
+		animate(src, 3 SECONDS, pixel_z = 20)
+		add_filter("flight_shadow", 2, drop_shadow_filter(y = 0, size = 0, color = "#000000"))
+		animate(get_filter("flight_shadow"), 3 SECONDS, y = -10, size = 10, flags = ANIMATION_PARALLEL)
+		coverage = 0
+		layer = FLY_LAYER
+		icon_state += "_flying"
+		for(var/mob/M in occupants)
+			M.set_flying(TRUE)
+		START_PROCESSING(SSobj, src)
+	else if(CHECK_BITFIELD(flags_pass, FLYING))
+		animate(src, 1 SECONDS, pixel_z = 0)
+		animate(get_filter("flight_shadow"), 1 SECONDS, y = 0, size = 0, flags = ANIMATION_PARALLEL)
+		STOP_PROCESSING(SSobj, src)
+		addtimer(CALLBACK(src, .proc/shutdown_procedures), 1 SECONDS)
+	else
+		balloon_alert(pilot, "Out of fuel!")
+
+///Series of actions taken when the aircraft is no longer flying
+/obj/vehicle/sealed/aircraft/proc/shutdown_procedures()
+	set_flying(FALSE)
+	coverage = initial(coverage)
+	layer = initial(layer)
+	remove_filter("flight_shadow")
+	icon_state = initial(icon_state)
+	for(var/mob/M in occupants)
+		M.set_flying(FALSE)
+	if(pilot)
+		pilot.update_action_buttons()	//For updating the toggle engine button
+	var/turf/landing_zone = get_turf(src)
+	if(landing_zone.density)
+		var/no_valid_turf = TRUE
+		for(var/turf/target in get_adjacent_open_turfs(src))
+			if(!is_blocked_turf(target))	//Try to find a valid open turf
+				throw_at(target, 2, 3, spin = TRUE)	//Throw the vehicle at this open turf
+				take_damage(FLYING_CRASH_DAMAGE)
+				no_valid_turf = FALSE
+				break
+		if(no_valid_turf)
+			obj_destruction()	//If no valid open turfs nearby, aircraft goes up in flames
+	if(LAZYLEN(landing_zone.contents))	//Let's crush anything this lands on!
+		for(var/atom/A in landing_zone.contents)
+			if(A == src)	//Don't splat ourselves!
+				continue
```

Comment:
stop using lazylen when its just to iterate through anyway 
iterating through null or an empty list does nothing anyway